lib/Net/SAML2/IdP.pm

```diff
@@ -223,22 +227,42 @@ Called after the object is created to validate the IdP using the cacert
 
 =cut
 
-sub BUILD {
-    my($self) = @_;
-
-    if ($self->cacert) {
-        my $ca = Crypt::OpenSSL::Verify->new($self->cacert, { strict_certs => 0, });
-
-        for my $use (keys %{$self->certs}) {
-            my $cert = Crypt::OpenSSL::X509->new_from_string($self->certs->{$use});
-            ## BUGBUG this is failing for valid things ...
-            eval { $ca->verify($cert) };
-            if ($@) {
-                warn "Can't verify IdP '$use' cert: $@\n";
+around BUILDARGS => sub {
+    my $orig = shift;
+    my $self = shift;
+
+    my %params = @_;
+
+    if ($params{cacert}) {
+        my $ca = Crypt::OpenSSL::Verify->new($params{cacert}, { strict_certs => 0, });
+
+        my $verified = 0;
+        my $error = "";
+        my @certs;
+
+        for my $pem (@{ $params{certs} }) {
+            for my $use (keys %{$pem}) {
+                my $cert = Crypt::OpenSSL::X509->new_from_string($pem->{$use});
+                ## BUGBUG this is failing for valid things ...
+                eval { $ca->verify($cert) };
+                if (!$@) {
```

Comment:
```
if ($@) { do the error condition; next }
# continue with the good stuff.
```
I think that we should work only with the good certs, ignore the incorrect ones and if that means that the XML cannot be validated for the incorrect certs, that will be noticed soon enough.

Now, if we have no correct certs: we should die, although I may opt for croaking. We don't do anything wrong, we are getting incorrect data from the caller.